README.md
<p align="center">
  <a href="https://goreportcard.com/report/github.com/portapps/slack-portable"><img src="https://goreportcard.com/badge/github.com/portapps/slack-portable?style=flat-square" alt="Go Report"></a>
  <a href="https://www.codacy.com/app/portapps/slack-portable"><img src="https://img.shields.io/codacy/grade/8beee2b3463842f6ad27da362666e75c.svg?style=flat-square" alt="Code Quality"></a>
  <a href="https://github.com/sponsors/crazy-max"><img src="https://img.shields.io/badge/sponsor-crazy--max-181717.svg?logo=github&style=flat-square" alt="Become a sponsor"></a>
  <a href="https://www.paypal.me/crazyws"><img src="https://img.shields.io/badge/donate-paypal-00457c.svg?logo=paypal&style=flat-square" alt="Donate Paypal"></a>
</p>

## Notice of Non-Affiliation and Disclaimer

Portapps is not affiliated, associated, authorized, endorsed by, or in any way officially connected with Slack™, or any of its subsidiaries or its affiliates.

The official Slack™ website can be found at https://slack.com.

The name Slack™ as well as related names, marks, emblems and images are registered trademarks of their respective owners.

## About

Slack™ portable app made with 🚀 [Portapps](https://portapps.io).<br />
Documentation and downloads can be found on https://portapps.io/app/slack-portable/

## How can I help ?

All kinds of contributions are welcome :raised_hands:! The most basic way to show your support is to star :star2: the project, or to raise issues :speech_balloon: You can also support this project by [**becoming a sponsor on GitHub**](https://github.com/sponsors/crazy-max) :clap: or by making a [Paypal donation](https://www.paypal.me/crazyws) to ensure this journey continues indefinitely! :rocket:

Thanks again for your support, it is much appreciated! :pray:

## License

MIT. See `LICENSE` for more details.<br />
Rocket icon credit to [Squid Ink](http://thesquid.ink).


CHANGELOG.md
# Changelog

## 4.23.0-36 (2022/01/04)

* Slack 4.23.0
* Portapps v3.4.0


## 4.3.4-35 (2020/02/12)

* Slack 4.3.4

## 4.3.2-34 (2020/01/23)

* Slack 4.3.2

## 4.3.0-33 (2020/01/03)

* Fix deep linking

## 4.3.0-32 (2019/12/16)

* Slack 4.3.0
* Add `cleanup` config
* Portapps 1.31.0

## 4.2.0-31 (2019/11/26)

* Slack 4.2.0
* Portapps 1.30.1

## 4.1.2-30 (2019/10/31)

* Slack 4.1.2

## 4.1.1-29 (2019/10/10)

* Slack 4.1.1
* Fix issue with deep linking

## 4.1.0-28 (2019/10/07)

* Fix deep linking
* Portapps 1.28.0

## 4.1.0-27 (2019/10/04)

* Slack 4.1.0
* Slack is now based on Electron 6
* Portapps 1.27.0

## 4.0.2-26 (2019/08/20)

* Slack 4.0.2
* Portapps 1.25.0

## 4.0.1-25 (2019/07/25)

* Slack 4.0.1

## 4.0.0-24 (2019/07/11)

* Slack 4.0.0

## 3.4.3-23 (2019/07/02)

* Slack 3.4.3

## 3.4.2-22 (2019/05/25)

* Slack 3.4.2
* Portapps 1.24.1

## 3.4.1-21 (2019/05/04)

* Slack 3.4.1
* Portapps 1.22.2

## 3.4.0-20 (2019/04/19)

* Slack 3.4.0
* Portapps 1.22.1

## 3.3.8-18 (2019/03/19)

* Create data folder at first launch
* Portapps 1.20.3

## 3.3.8-17 (2019/03/09)

* Slack 3.3.8
* Portapps 1.20.2

## 3.3.7-41 (2019/01/18)

* Slack 3.3.7
* [New Slack logo](https://slackhq.com/say-hello-new-logo)

## 3.3.3-40 (2018/10/24)

* Fix Windows desktop notifications

## 3.3.3-39 (2018/10/02)

* Slack 3.3.3
* Go 1.11
* Use [go mod](https://golang.org/cmd/go/#hdr-Module_maintenance) instead of `dep`

## 3.3.1-38 (2018/09/01)

* Slack 3.3.1

## 3.3.0-37 (2018/08/15)

* Slack 3.3.0

## 3.2.0-36 (2018/06/06)

* Slack 3.2.0

## 3.1.1-35 (2018/04/05)

* Slack 3.1.1

## 3.1.0-34 (2018/03/10)

* Slack 3.1.0

## 3.0.5-33 (2018/02/11)

* Move ia32/x64 to win32/win64 for arch def
* Remove nupkg file
* Add portapp.json file
* Uncheck run app in setup

## 3.0.5-32 (2018/02/09)

* Ability to pass custom args to the portable process

## 3.0.5-31 (2018/01/21)

* Slack 3.0.5

## 3.0.3-30 (2018/01/13)

* Rebuild electron.asar to push data directly in `data` subfolder
* No need to override USERPROFILE environment variable anymore

> :warning: **UPGRADE NOTES**
> * Move everything in `data\AppData\Roaming\slack\*` to `data` folder and remove folder `data\AppData`.

## 3.0.3-29 (2018/01/12)

* Slack 3.0.3

## 3.0.0-28 (2017/12/09)

* Slack 3.0.0

## 2.9.0-27 (2017/11/23)

* Move app to a subfolder
* Reduce dependencies to avoid heuristic detection
* Add UPX compression

> :warning: **UPGRADE NOTES**
> * Delete all files and folders in root folder except `data` folder.

## 2.9.0-26 (2017/11/19)

* Slack 2.9.0
* Move repository to [Portapps](https://github.com/portapps) org
* Switch to [Golang Dep](https://github.com/golang/dep) as dependency manager
* Go 1.9.1

## 2.8.2-25 (2017/10/19)

* Slack 2.8.2

## 2.8.1-24 (2017/10/01)

* Slack 2.8.1

## 2.8.0-23 (2017/09/14)

* Slack 2.8.0
* New instance opened with Windows notification method (Issue #8)

## 2.7.1-22 (2017/09/05)

* New logger
* Override USERPROFILE env var instead of using symlink to APPDATA to store data
* Do not migrate old data folder from APPDATA
* Reduce dependencies and system calls to avoid heuristic detection

> :warning: **UPGRADE NOTES**
> * Move the content of `data\*` in `data\AppData\Roaming\slack\`
> * Remove symlink `%APPDATA%\slack`

## 2.7.1-21 (2017/08/26)

* Go 1.9
* Add support guidelines

## 2.7.1-19 (2017/08/16)

* Slack 2.7.1

## 2.7.0-18 (2017/08/09)

* Slack 2.7.0
* ia32 releases now included
* Small refactoring

## 2.6.5-16 (2017/07/13)

* Slack 2.6.5

## 2.6.3-15 (2017/06/29)

* Slack 2.6.3
* Admin privileges not required for Setup

## 2.6.2-14 (2017/05/24)

* Slack 2.6.2
* Download Glide as an external lib

## 2.6.0-beta189985592-13 (2017/05/14)

* Provide the nupkg file in the release
* Add Go report card 
* Add Glide dependency manager for Go
* Standard code organization

## 2.6.0-beta189985592-12 (2017/05/03)

* Error on first install with InnoSetup (Issue #7)

## 2.6.0-beta189985592-10 (2017/04/28)

* Create a setup for portable installation (Issue #6)

## 2.6.0-beta189985592-9 (2017/04/28)

* Add Slack logging (Issue #4)
* Add ability to select the Slack version (Issue #3)
* Slack crashes because of Update.exe missing (Issue #2)

## 2.6.0-beta189985592-7 (2017/04/27)

* Not launched if data folder does not exists (Issue #1)
* Improve logging
* Slack 2.6.0-beta189985592

## 2.5.2-5 (2017/04/10)

* GetFileAttributesEx error when no data folder exists

## 2.5.2-1 (2017/04/09)

* Initial version


main.go
//go:generate go install -v github.com/kevinburke/go-bindata/go-bindata
//go:generate go-bindata -prefix res/ -pkg assets -o assets/assets.go res/Slack.lnk
//go:generate go install -v github.com/josephspurrier/goversioninfo/cmd/goversioninfo
//go:generate goversioninfo -icon=res/papp.ico -manifest=res/papp.manifest
package main

import (
	"encoding/json"
	"io/ioutil"
	"os"
	"path"

	. "github.com/portapps/portapps"
	"github.com/portapps/portapps/pkg/shortcut"
	"github.com/portapps/portapps/pkg/utl"
	"github.com/portapps/slack-portable/assets"
)

type config struct {
	Cleanup bool `yaml:"cleanup" mapstructure:"cleanup"`
}

var (
	app *App
	cfg *config
)

func init() {
	var err error

	// Default config
	cfg = &config{
		Cleanup: false,
	}

	// Init app
	if app, err = NewWithCfg("slack-portable", "Slack", cfg); err != nil {
		Log.Fatal().Err(err).Msg("Cannot initialize application. See log file for more info.")
	}
}

func main() {
	utl.CreateFolder(app.DataPath)
	electronBinPath := utl.PathJoin(app.AppPath, utl.FindElectronAppFolder("app-", app.AppPath))

	app.Process = utl.PathJoin(electronBinPath, "Slack.exe")
	app.Args = []string{
		"--user-data-dir=" + app.DataPath,
	}
	app.WorkingDir = electronBinPath

	// Cleanup on exit
	if cfg.Cleanup {
		defer func() {
			utl.Cleanup([]string{
				path.Join(os.Getenv("APPDATA"), "Slack"),
			})
		}()
	}

	// Downloads folder
	downloadsPath := utl.CreateFolder(app.RootPath, "downloads")

	// Update slack settings
	Log.Info().Msg("Update Slack settings...")
	slackSettingsPath := utl.PathJoin(app.DataPath, "storage", "slack-settings")
	if _, err := os.Stat(slackSettingsPath); err == nil {
		rawSettings, err := ioutil.ReadFile(slackSettingsPath)
		if err == nil {
			jsonMapSettings := make(map[string]interface{})
			if err = json.Unmarshal(rawSettings, &jsonMapSettings); err != nil {
				Log.Error().Err(err).Msg("Settings unmarshal")
			}
			Log.Info().Interface("settings", jsonMapSettings).Msg("Current settings")

			jsonMapSettings["resourcePath"] = utl.PathJoin(electronBinPath, "resources", "app.asar")
			jsonMapSettings["PrefSSBFileDownloadPath"] = downloadsPath
			jsonMapSettings["notificationMethod"] = "html"
			Log.Info().Interface("settings", jsonMapSettings).Msg("New settings")

			jsonSettings, err := json.Marshal(jsonMapSettings)
			if err != nil {
				Log.Error().Err(err).Msg("Settings marshal")
			}
			err = ioutil.WriteFile(slackSettingsPath, jsonSettings, 0644)
			if err != nil {
				Log.Error().Err(err).Msg("Write settings")
			}
		}
	} else {
		Log.Error().Msgf("Slack settings not found in %s", slackSettingsPath)
	}

	// Copy default shortcut
	shortcutPath := path.Join(utl.StartMenuPath(), "Slack Portable.lnk")
	defaultShortcut, err := assets.Asset("Slack.lnk")
	if err != nil {
		Log.Error().Err(err).Msg("Cannot load asset Slack.lnk")
	}
	err = ioutil.WriteFile(shortcutPath, defaultShortcut, 0644)
	if err != nil {
		Log.Error().Err(err).Msg("Cannot write default shortcut")
	}

	// Update default shortcut
	err = shortcut.Create(shortcut.Shortcut{
		ShortcutPath:     shortcutPath,
		TargetPath:       app.Process,
		Arguments:        shortcut.Property{Clear: true},
		Description:      shortcut.Property{Value: "Slack Portable by Portapps"},
		IconLocation:     shortcut.Property{Value: app.Process},
		WorkingDirectory: shortcut.Property{Value: app.AppPath},
	})
	if err != nil {
		Log.Error().Err(err).Msg("Cannot create shortcut")
	}
	defer func() {
		if err := os.Remove(shortcutPath); err != nil {
			Log.Error().Err(err).Msg("Cannot remove shortcut")
		}
	}()

	// Disable auto updates
	utl.OverrideEnv("SLACK_NO_AUTO_UPDATES", "true")

	// Update deep link
	mainJs := utl.PathJoin(electronBinPath, "resources", "app", "dist", "main.bundle.js")
	contains, err := utl.FileContains(mainJs, `require('./portapps.js');`)
	if !contains && err == nil {
		Log.Info().Msgf("Updating content of %s", mainJs)
		if err := utl.AppendToFile(mainJs, ` require('./portapps.js');`); err != nil {
			Log.Error().Err(err).Msgf("Cannot append content to %s", mainJs)
		}
	} else if err != nil {
		Log.Error().Err(err).Msgf("Cannot check %s", mainJs)
	}
	err = utl.WriteToFile(utl.PathJoin(electronBinPath, "resources", "app", "dist", "portapps.js"), `"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const path = require("path");
const { app } = require('electron');
app.on('browser-window-created', () => {
    if (app.listenerCount('browser-window-created') <= 10) {
        app.setAsDefaultProtocolClient('slack', process.execPath, ["--user-data-dir="+path.join(path.dirname(process.execPath), '..', '..', 'data')]);
    }
});`)
	if err != nil {
		Log.Error().Err(err).Msg("Cannot write to portapps.js")
	}

	app.Launch(os.Args[1:])
}


assets/assets.go
// Code generated by go-bindata. DO NOT EDIT.
// sources:
// res/Slack.lnk (1.798kB)

package assets

import (
	"bytes"
	"compress/gzip"
	"crypto/sha256"
	"fmt"
	"io"
	"io/ioutil"
	"os"
	"path/filepath"
	"strings"
	"time"
)

func bindataRead(data []byte, name string) ([]byte, error) {
	gz, err := gzip.NewReader(bytes.NewBuffer(data))
	if err != nil {
		return nil, fmt.Errorf("read %q: %v", name, err)
	}

	var buf bytes.Buffer
	_, err = io.Copy(&buf, gz)
	clErr := gz.Close()

	if err != nil {
		return nil, fmt.Errorf("read %q: %v", name, err)
	}
	if clErr != nil {
		return nil, err
	}

	return buf.Bytes(), nil
}

type asset struct {
	bytes  []byte
	info   os.FileInfo
	digest [sha256.Size]byte
}

type bindataFileInfo struct {
	name    string
	size    int64
	mode    os.FileMode
	modTime time.Time
}

func (fi bindataFileInfo) Name() string {
	return fi.name
}
func (fi bindataFileInfo) Size() int64 {
	return fi.size
}
func (fi bindataFileInfo) Mode() os.FileMode {
	return fi.mode
}
func (fi bindataFileInfo) ModTime() time.Time {
	return fi.modTime
}
func (fi bindataFileInfo) IsDir() bool {
	return false
}
func (fi bindataFileInfo) Sys() interface{} {
	return nil
}

var _slackLnk = []byte("\x1f\x8b\x08\x00\x00\x00\x00\x00\x00\xff\xec\x94\x4f\x68\x2b\x55\x14\xc6\x7f\x69\xa3\xb1\xa2\x90\x6a\x10\x5b\x2a\xcd\x26\x22\xca\x8c\xc9\x34\x6d\x32\x41\xa1\x25\x7f\x48\xe9\xd0\x84\xc4\x40\x91\x11\xbc\x4e\xa7\x34\x34\x21\x63\x12\xa1\x75\x21\xb8\x2a\x96\xba\xd1\x45\x2b\x28\xad\x22\xb4\x82\x06\xeb\xb6\x52\x5c\x5a\x17\xd2\x45\xa5\x6a\x41\xad\xe8\x42\x41\x0b\xba\x12\x5c\xc8\xdc\x24\x36\xed\x2b\x3c\x78\xbb\xf7\x78\x1f\xcc\x9c\xb9\xe7\xcf\x77\xbe\x3b\x97\x7b\x0c\xc0\x13\xe8\xc3\xc5\x17\xf2\x4d\xe6\x64\x12\x82\xc0\xf9\xa7\x7d\x83\x87\xad\x63\xcf\xbd\x27\x6d\x2b\x76\xf6\x7f\xf9\xb2\x75\xec\xf1\x2f\x78\x65\xa2\x87\xcb\x38\x25\xc0\x68\xfe\xa7\xdc\x51\xf0\xf7\x44\xd9\xff\xc1\xb7\xf7\xf1\x54\x38\xfc\xde\x10\x4f\x27\x13\x26\x37\x22\x4f\x44\x5a\x6d\xd6\xf1\xf9\x69\x54\x84\xb5\xc4\x33\x0c\xe0\xe5\xfc\x40\x9b\xad\xf9\x5c\xbf\x0a\x6c\x8e\xb9\xf2\xee\xb9\x52\x9d\x78\xe0\x4f\x1a\x54\x10\x58\x2c\x01\x01\x4c\x34\x5c\x69\xda\x6c\xd6\x17\x6c\xf3\xa9\xf6\xb2\x4d\xea\x12\x67\x4d\x72\xbe\x2f\x39\xbd\x57\x38\xb7\xde\xcd\xf5\x70\xaa\xd8\x2c\x63\x03\x8f\x02\x53\xc0\x48\x67\xd3\xae\x55\x3a\x35\x93\xc0\x20\xd0\x0f\x24\xf7\x47\x4c\xbf\xeb\x4c\x26\x4c\xd9\xdf\xbc\x50\xc1\x83\x14\xff\xa7\x0e\x92\xc2\xa6\xc1\x12\x4d\x6a\x38\x3c\x44\x92\x04\x66\x4f\x73\x13\x81\x83\x83\xc2\x18\xaa\x7c\x62\x3c\xd9\xc9\x2a\xd1\xc0\xa6\x4e\x03\x13\x8b\x3a\x82\x57\x59\xc1\x64\x4a\x56\xa4\x10\x34\x11\x98\x18\xd4\xb0\x10\x54\xae\x65\x56\x29\x63\x51\x23\xd0\x0f\x3e\xd8\x0e\x95\x8a\xe9\x42\xbe\x90\xcb\x4c\x1b\xe9\x90\x39\xe5\x38\x29\xd1\x14\xa6\x51\xb3\x44\xa5\xb3\x19\xe1\x38\x6a\xd9\xaa\x5d\x73\x94\x77\x0a\x42\x94\x28\x92\xa6\x40\x9e\x02\x39\x32\x4c\x63\x90\x26\x74\xcb\x7f\xf7\x2e\x6e\x47\xbc\xd8\xbe\xd0\xdb\x73\x9d\xb5\x63\x29\x8b\x62\xde\x6e\x74\xe3\xf3\xad\x56\x73\xe5\xaf\x8f\xd3\x6f\xbd\xfe\xeb\x47\xdf\xef\xae\x7a\xbf\x7b\x7e\xe3\x70\x6f\xf8\xb7\xc1\x77\x8c\xe7\x3e\xeb\xbf\x7f\xcd\xb8\x59\x7c\xdd\x03\x03\xb0\xfd\x06\x10\x29\xe6\x8b\x3f\xaf\xcd\x65\x3e\x37\xe2\xc9\xfd\x7f\x1f\x7e\xfb\xf1\xcd\xea\xd7\xd5\xee\x70\x1a\x05\xdc\x79\x55\x44\x21\x82\xc2\x38\x0a\x9a\xfc\x1a\x23\x86\x4e\x94\x28\x3a\x3a\x13\xc4\x99\x90\x31\x9d\x71\xc2\x44\xe5\x5a\x23\x2e\x33\xc7\x89\x12\x26\x4c\x84\x98\xac\x98\x90\x6c\x6d\x4f\x1b\x07\x1d\x21\xa5\x27\x8c\xad\x95\x2d\x7d\x66\xf7\xe8\xec\x58\x39\xfb\xe6\xef\x17\x80\xe1\xae\x90\x10\xa0\x23\xd0\xd1\x98\x47\x10\x27\x86\x22\x05\x2c\xa0\x49\x71\x11\x04\x51\x14\x04\x1a\x2f\x49\xb1\x31\x22\xc4\x65\xd3\x28\x36\x0b\x58\x44\xa4\x1c\x17\xe9\xee\x58\x77\xc9\x87\x40\x5e\x98\x2a\x2a\x0d\x5e\xe6\x15\xca\xd4\xa9\x63\x53\x91\x9e\x8b\xa9\xdc\x3b\xf5\xdb\xd0\x3b\xea\xf7\x1e\xa9\xa6\x3e\x79\xd3\xc9\xee\x64\x27\xd5\x0f\x9f\x5d\x5e\x7f\x0c\x58\x74\x13\xb2\xc0\x8f\x3f\x0c\xff\xb1\xfa\xcf\x79\x76\xa3\x75\x3a\xe3\xfb\xea\xb5\x70\xef\x79\xff\x17\x00\x00\xff\xff\x71\x6c\x57\x8d\x06\x07\x00\x00")

func slackLnkBytes() ([]byte, error) {
	return bindataRead(
		_slackLnk,
		"Slack.lnk",
	)
}

func slackLnk() (*asset, error) {
	bytes, err := slackLnkBytes()
	if err != nil {
		return nil, err
	}

	info := bindataFileInfo{name: "Slack.lnk", size: 1798, mode: os.FileMode(0666), modTime: time.Unix(1547773182, 0)}
	a := &asset{bytes: bytes, info: info, digest: [32]uint8{0xc5, 0xed, 0x4a, 0x31, 0xef, 0x53, 0x4a, 0x6c, 0x45, 0x4c, 0x6e, 0xed, 0x5, 0x25, 0x87, 0xdc, 0xe3, 0xc, 0xda, 0xed, 0x9, 0x34, 0xcf, 0x33, 0xcd, 0x51, 0xb2, 0xc9, 0xb0, 0x97, 0x9b, 0xef}}
	return a, nil
}

// Asset loads and returns the asset for the given name.
// It returns an error if the asset could not be found or
// could not be loaded.
func Asset(name string) ([]byte, error) {
	canonicalName := strings.Replace(name, "\\", "/", -1)
	if f, ok := _bindata[canonicalName]; ok {
		a, err := f()
		if err != nil {
			return nil, fmt.Errorf("Asset %s can't read by error: %v", name, err)
		}
		return a.bytes, nil
	}
	return nil, fmt.Errorf("Asset %s not found", name)
}

// AssetString returns the asset contents as a string (instead of a []byte).
func AssetString(name string) (string, error) {
	data, err := Asset(name)
	return string(data), err
}

// MustAsset is like Asset but panics when Asset would return an error.
// It simplifies safe initialization of global variables.
func MustAsset(name string) []byte {
	a, err := Asset(name)
	if err != nil {
		panic("asset: Asset(" + name + "): " + err.Error())
	}

	return a
}

// MustAssetString is like AssetString but panics when Asset would return an
// error. It simplifies safe initialization of global variables.
func MustAssetString(name string) string {
	return string(MustAsset(name))
}

// AssetInfo loads and returns the asset info for the given name.
// It returns an error if the asset could not be found or
// could not be loaded.
func AssetInfo(name string) (os.FileInfo, error) {
	canonicalName := strings.Replace(name, "\\", "/", -1)
	if f, ok := _bindata[canonicalName]; ok {
		a, err := f()
		if err != nil {
			return nil, fmt.Errorf("AssetInfo %s can't read by error: %v", name, err)
		}
		return a.info, nil
	}
	return nil, fmt.Errorf("AssetInfo %s not found", name)
}

// AssetDigest returns the digest of the file with the given name. It returns an
// error if the asset could not be found or the digest could not be loaded.
func AssetDigest(name string) ([sha256.Size]byte, error) {
	canonicalName := strings.Replace(name, "\\", "/", -1)
	if f, ok := _bindata[canonicalName]; ok {
		a, err := f()
		if err != nil {
			return [sha256.Size]byte{}, fmt.Errorf("AssetDigest %s can't read by error: %v", name, err)
		}
		return a.digest, nil
	}
	return [sha256.Size]byte{}, fmt.Errorf("AssetDigest %s not found", name)
}

// Digests returns a map of all known files and their checksums.
func Digests() (map[string][sha256.Size]byte, error) {
	mp := make(map[string][sha256.Size]byte, len(_bindata))
	for name := range _bindata {
		a, err := _bindata[name]()
		if err != nil {
			return nil, err
		}
		mp[name] = a.digest
	}
	return mp, nil
}

// AssetNames returns the names of the assets.
func AssetNames() []string {
	names := make([]string, 0, len(_bindata))
	for name := range _bindata {
		names = append(names, name)
	}
	return names
}

// _bindata is a table, holding each asset generator, mapped to its name.
var _bindata = map[string]func() (*asset, error){
	"Slack.lnk": slackLnk,
}

// AssetDir returns the file names below a certain
// directory embedded in the file by go-bindata.
// For example if you run go-bindata on data/... and data contains the
// following hierarchy:
//     data/
//       foo.txt
//       img/
//         a.png
//         b.png
// then AssetDir("data") would return []string{"foo.txt", "img"},
// AssetDir("data/img") would return []string{"a.png", "b.png"},
// AssetDir("foo.txt") and AssetDir("notexist") would return an error, and
// AssetDir("") will return []string{"data"}.
func AssetDir(name string) ([]string, error) {
	node := _bintree
	if len(name) != 0 {
		canonicalName := strings.Replace(name, "\\", "/", -1)
		pathList := strings.Split(canonicalName, "/")
		for _, p := range pathList {
			node = node.Children[p]
			if node == nil {
				return nil, fmt.Errorf("Asset %s not found", name)
			}
		}
	}
	if node.Func != nil {
		return nil, fmt.Errorf("Asset %s not found", name)
	}
	rv := make([]string, 0, len(node.Children))
	for childName := range node.Children {
		rv = append(rv, childName)
	}
	return rv, nil
}

type bintree struct {
	Func     func() (*asset, error)
	Children map[string]*bintree
}

var _bintree = &bintree{nil, map[string]*bintree{
	"Slack.lnk": &bintree{slackLnk, map[string]*bintree{}},
}}

// RestoreAsset restores an asset under the given directory.
func RestoreAsset(dir, name string) error {
	data, err := Asset(name)
	if err != nil {
		return err
	}
	info, err := AssetInfo(name)
	if err != nil {
		return err
	}
	err = os.MkdirAll(_filePath(dir, filepath.Dir(name)), os.FileMode(0755))
	if err != nil {
		return err
	}
	err = ioutil.WriteFile(_filePath(dir, name), data, info.Mode())
	if err != nil {
		return err
	}
	return os.Chtimes(_filePath(dir, name), info.ModTime(), info.ModTime())
}

// RestoreAssets restores an asset under the given directory recursively.
func RestoreAssets(dir, name string) error {
	children, err := AssetDir(name)
	// File
	if err != nil {
		return RestoreAsset(dir, name)
	}
	// Dir
	for _, child := range children {
		err = RestoreAssets(dir, filepath.Join(name, child))
		if err != nil {
			return err
		}
	}
	return nil
}

func _filePath(dir, name string) string {
	canonicalName := strings.Replace(name, "\\", "/", -1)
	return filepath.Join(append([]string{dir}, strings.Split(canonicalName, "/")...)...)
}


slack-portable.go
//go:generate go get -v github.com/josephspurrier/goversioninfo/...
//go:generate goversioninfo -icon=slack-portable.ico
package main

import (
	"encoding/json"
	"fmt"
	"io"
	"io/ioutil"
	"os"
	"os/exec"
	"path"
	"path/filepath"
	"strings"
	"syscall"

	"github.com/op/go-logging"
)

var log = logging.MustGetLogger("slack-portable")
var logFormat = logging.MustStringFormatter(`%{time:2006-01-02 15:04:05} %{level:.4s} - %{message}`)

func main() {
	// Current path
	currentPath, err := filepath.Abs(filepath.Dir(os.Args[0]))
	if err != nil {
		log.Error("Current path:", err)
	}

	// Logs folder
	var logsPath = path.Join(currentPath, "logs")
	if _, err := os.Stat(logsPath); os.IsNotExist(err) {
		log.Info("Create logs folder", logsPath)
		err = os.Mkdir(logsPath, 777)
		if err != nil {
			log.Error("Create logs folder:", err)
		}
	}

	// Old behaviors
	var oldLog = path.Join(currentPath, "slack-portable.log")
	if _, err := os.Stat(oldLog); err == nil {
		err = os.Remove(oldLog)
		if err != nil {
			log.Error("Remove old log:", err)
		}
	}

	// Log file
	logfile, err := os.OpenFile(path.Join(logsPath, "slack-portable.log"), os.O_WRONLY|os.O_CREATE|os.O_APPEND, 0644)
	if err != nil {
		log.Error("Log file:", err)
	}

	// Init logger
	logBackendStdout := logging.NewBackendFormatter(logging.NewLogBackend(os.Stdout, "", 0), logFormat)
	logBackendFile := logging.NewBackendFormatter(logging.NewLogBackend(logfile, "", 0), logFormat)
	logging.SetBackend(logBackendStdout, logBackendFile)
	log.Info("--------")
	log.Info("Starting slack-portable...")

	// Purge logs
	logsFolder, err := os.Open(logsPath)
	if err != nil {
		log.Error("Open logs folder:", err)
	}
	defer logsFolder.Close()
	logsFiles, err := logsFolder.Readdir(-1)
	if err != nil {
		log.Error("Read logs folder:", err)
	}
	log.Info("Reading", logsPath)
	for _, logsFile := range logsFiles {
		if !strings.HasPrefix(logsFile.Name(), "slack-portable") {
			os.Remove(path.Join(logsPath, logsFile.Name()))
			log.Info("Deleted", path.Join(logsPath, logsFile.Name()))
		}
	}

	// Convert backslashes
	currentPath = path.Join(strings.Replace(string(currentPath), string(filepath.Separator), "/", -1))
	log.Info("Current path:" + currentPath)

	// Find app folder
	log.Info("Lookup app folder in", currentPath)
	var appPath = ""
	rootFiles, _ := ioutil.ReadDir(currentPath)
	for _, f := range rootFiles {
		if strings.HasPrefix(f.Name(), "app-") && f.IsDir() {
			log.Info("App folder found:", f.Name())
			appPath = path.Join(currentPath, f.Name())
			break
		}
	}
	if _, err := os.Stat(appPath); err == nil {
		log.Info("App path:", appPath)
	} else {
		log.Error("App path does not exist")
	}

	// Init vars
	var slackExe = path.Join(appPath, "slack.exe")
	var dataPath = path.Join(currentPath, "data")
	var downloadsPath = path.Join(currentPath, "downloads")
	var symlinkPath = path.Clean(path.Join(os.Getenv("APPDATA"), "Slack"))
	//var symlinkPath = path.Join(currentPath, "data2")
	var slackSettingsPath = path.Join(dataPath, "storage", "slack-settings")
	log.Info("Slack executable:", slackExe)
	log.Info("Data path:", dataPath)
	log.Info("Downloads path:", downloadsPath)
	log.Info("Symlink path:", symlinkPath)
	log.Info("Slack settings path:", slackSettingsPath)

	// Create data folder
	if _, err := os.Stat(dataPath); os.IsNotExist(err) {
		log.Info("Create data folder...", dataPath)
		err = os.Mkdir(dataPath, 777)
		if err != nil {
			log.Error("Create data folder:", err)
		}
	}

	// Check old data folder
	if _, err := os.Stat(symlinkPath); err == nil {
		fi, err := os.Lstat(symlinkPath)
		if err != nil {
			log.Error("Symlink lstat:", err)
		}
		if fi.Mode()&os.ModeSymlink != os.ModeSymlink {
			// Copy old data folder
			log.Info("Copy old data from", symlinkPath)
			err = copyDir(symlinkPath, dataPath)
			if err != nil {
				log.Error("Copying old data folder:", err)
			}

			// Rename old data folder
			log.Info("Chmod old data folder...")
			err = os.Chmod(symlinkPath, 0777)
			if err != nil {
				log.Error("Chmod old data folder:", err)
			}

			log.Info("Rename old data folder to", symlinkPath+"_old")
			err = os.Rename(symlinkPath, symlinkPath+"_old")
			if err != nil {
				log.Error("Renaming old data folder:", err)
			}
		}
	}

	// Create symlink
	log.Info("Create symlink", symlinkPath)
	os.Remove(symlinkPath)
	cmd := exec.Command("cmd", "/c", "mklink", "/J", strings.Replace(symlinkPath, "/", "\\", -1), strings.Replace(dataPath, "/", "\\", -1))
	cmd.SysProcAttr = &syscall.SysProcAttr{HideWindow: true}
	if err := cmd.Run(); err != nil {
		log.Error("Symlink:", err)
	}
	/*err = os.Symlink(dataPath, symlinkPath)
	  if err != nil {
	    log.Error(err)
	  }*/

	// Check downloads folder
	log.Info("Check downloads folder...")
	if _, err := os.Stat(downloadsPath); os.IsNotExist(err) {
		log.Info("Create download folder", downloadsPath)
		err = os.Mkdir(downloadsPath, 777)
		if err != nil {
			log.Error("Create download folder:", err)
		}
	}

	// Change slack settings
	log.Info("Update Slack settings...")
	if _, err := os.Stat(slackSettingsPath); err == nil {
		rawSettings, err := ioutil.ReadFile(slackSettingsPath)
		if err == nil {
			jsonMapSettings := make(map[string]interface{})
			json.Unmarshal(rawSettings, &jsonMapSettings)
			log.Info("Current settings:", jsonMapSettings)

			jsonMapSettings["resourcePath"] = strings.Replace(appPath+"/resources/app.asar", "/", "\\", -1)
			jsonMapSettings["PrefSSBFileDownloadPath"] = strings.Replace(downloadsPath, "/", "\\", -1)
			log.Info("New settings:", jsonMapSettings)

			jsonSettings, err := json.Marshal(jsonMapSettings)
			if err != nil {
				log.Error("Slack settings marshal:", err)
			}
			err = ioutil.WriteFile(slackSettingsPath, jsonSettings, 0644)
			if err != nil {
				log.Error("Write Slack settings:", err)
			}
		}
	} else {
		log.Warning("Slack settings not found in:", slackSettingsPath)
	}

	// Launch slack
	log.Info("Launch Slack...")
	cmd = exec.Command(slackExe, "--log-file", "./")
	cmd.Dir = logsPath

	defer logfile.Close()
	cmd.Stdout = logfile
	cmd.Stderr = logfile

	if err := cmd.Start(); err != nil {
		log.Error("Cmd Start:", err)
	}

	cmd.Wait()
}

// src: https://gist.github.com/m4ng0squ4sh/92462b38df26839a3ca324697c8cba04
func copyFile(src, dst string) (err error) {
	in, err := os.Open(src)
	if err != nil {
		return
	}
	defer in.Close()

	out, err := os.Create(dst)
	if err != nil {
		return
	}
	defer func() {
		if e := out.Close(); e != nil {
			err = e
		}
	}()

	_, err = io.Copy(out, in)
	if err != nil {
		return
	}

	err = out.Sync()
	if err != nil {
		return
	}

	si, err := os.Stat(src)
	if err != nil {
		return
	}

	err = os.Chmod(dst, si.Mode())
	if err != nil {
		return
	}

	return
}

// src: https://gist.github.com/m4ng0squ4sh/92462b38df26839a3ca324697c8cba04
func copyDir(src string, dst string) (err error) {
	src = filepath.Clean(src)
	dst = filepath.Clean(dst)

	si, err := os.Stat(src)
	if err != nil {
		return err
	}
	if !si.IsDir() {
		return fmt.Errorf("source is not a directory")
	}

	_, err = os.Stat(dst)
	if err != nil && !os.IsNotExist(err) {
		return
	}

	err = os.MkdirAll(dst, si.Mode())
	if err != nil {
		return
	}

	entries, err := ioutil.ReadDir(src)
	if err != nil {
		return
	}

	for _, entry := range entries {
		srcPath := filepath.Join(src, entry.Name())
		dstPath := filepath.Join(dst, entry.Name())
		if entry.IsDir() {
			err = copyDir(srcPath, dstPath)
			if err != nil {
				return
			}
		} else {
			err = copyFile(srcPath, dstPath)
			if err != nil {
				return
			}
		}
	}

	return
}
